### PR TITLE
Add `bless` x.py subcommand for easy ui test replacement

### LINKED ISSUE #49815: move the `update-reference.sh` script into `x.py`
The script that updates reference files for UI tests is currently a shell script, which has portability issues. It is also a bit annoying to use since you must specify the path to your build directory and so forth. It'd be nice to fix all these problems and move it into `x.py`.

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #49815

r? @nikomatsakis

### CODE REVIEW

**Comment by oli-obk** [comment]
cc @ehuss

**Comment by kennytm** [comment]
With the implementation in this PR, I think it should be a `--bless` flag to `test` instead of an entire new subcommand.

```console
$ ./x.py test src/test/ui --bless
```

In  #49815's suggestion the subcommand works like `update-reference.sh`. These scripts won't run tests themselves, they only copy the generated stdout/stderr files, meaning the workflow would be like

```console
$ ./x.py test src/test/ui

(... fix errors ...)

$ ./x.py test src/test/ui

(... fix errors ...)

$ ./x.py test src/test/ui

(done)

$ ./x.py bless src/test/ui # <- will not run tests again
```

**Comment by petrochenkov** [comment]
In my experience the workflow is
```
// 1. Initial run
x.py test ui
// 2. Update stderr/stdout
update-all-reference.sh
// 3. Was it enough or some `//~ ERROR` annotations need update as well?
x.py test ui
// 4. Optionally: Fix `//~ ERROR` annotations
(... fix errors ...)
// 5. Was it enough or stderr/stdout need updating again?
x.py test ui
// 6. Optionally: Update stderr/stdout
update-all-reference.sh
// 7. Make sure everything passes now, maybe go back to `4.` if not
x.py test ui
```
so I'd very much prefer to do the "test-update-test" operation in one command to keeping "update" logic separate.
Then the workflow would turn into
```
// 1. Initial "test-update-test" run
x.py bless ui
// 2. Optionally: Fix `//~ ERROR` annotations
(... fix errors ...)
// 3. Second "test-update-test" run to make sure everything passes now, maybe go back to `2.` if not
x.py bless ui
```

I'm not sure about `test --bless` vs `bless`, the first one is more correct because it's "`test` with a tweak", but the second one is more convenient.

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -2979,19 +2981,42 @@ impl<'test> TestCx<'test> {
...
-            Ok(()) => {}
-            Err(e) => self.fatal(&format!(
-                "failed to write {} to `{}`: {}",
-                kind,
-                output_file.display(),
-                e
-            )),
+        let mut files = vec![output_file];
+        if self.config.bless {
+            files.push(self.expected_output_path(kind));
```
Comment (ehuss) [inline src/tools/compiletest/src/runtest.rs]: This doesn't seem to work for `nll` files.  For an example, copy E0508.rs to `src/test/ui` and try to bless it.  You should end up with these three files:
```
E0508.ast.nll.stderr
E0508.ast.stderr
E0508.mir.stderr
```

I think the issue is that `expected_output_path` doesn't include mode if the file doesn't already exist.  I think the solution is to call the global `expected_output_path` (not the `self` version).

**Review by ehuss — COMMENTED** [review]
Sweet!  Sorry I didn't get to this sooner!  (BTW, love the branch name.)

Maybe the error message that says "To update references, run this command..." should be updated to the new command?

Also, beware that there are a bunch of path changes coming in #50400.  There shouldn't be any fundamental problems, but some of the function names will change.

**Comment by nikomatsakis** [comment]
@ehuss -- you're the one most familiar with the code at this point, would you be up for tackling the review here? (I can delegate to you from bors)

**Comment by ehuss** [comment]
@nikomatsakis I don't mind, but I don't have any more comments beyond what I posted.

**Comment by killercup** [comment]
@oli-obk want to add replacement of `.fixed` files, too? (not necessarily in this pr though)

**Comment by nikomatsakis** [comment]
@oli-obk are you ready for a re-review from @ehuss then?

**Comment by oli-obk** [comment]
Yea, I was just waiting for travis :) Last time I broke some tests that ran very late in the pipeline.

**Comment by ehuss** [comment]
@bors: r+

Nice!

**Pull request merged**